src/shred_collector/shred_receiver.zig

```diff
@@ -80,76 +80,69 @@ pub const ShredReceiver = struct {
-                            try self.handlePacket(packet, &responses, shred_version);
-                        }
-                        try self.unverified_shred_sender.send(batch);
-                    }
+                while (receiver.receive()) |packet| {
+                    var our_packet = packet;
+                    try self.handlePacket(&our_packet, &responses);
+                    if (is_repair) our_packet.flags.set(.repair);
+                    try self.unverified_shred_sender.send(our_packet);
                 } else {
                     std.time.sleep(10 * std.time.ns_per_ms);
                 }
             }
-            if (responses.items.len > 0) {
-                try response_sender.send(responses);
+            for (responses.items) |response| {
+                try response_sender.send(response);
             }
         }
     }
 
     /// Handle a single packet and return
     fn handlePacket(
-        self: *Self,
+        self: Self,
         packet: *Packet,
         responses: *ArrayList(Packet),
-        shred_version: u16,
     ) !void {
         if (packet.size == REPAIR_RESPONSE_SERIALIZED_PING_BYTES) {
-            try self.handlePing(packet, responses);
-            packet.flags.set(.discard);
+            if (try self.handlePing(packet)) |p| try responses.append(p);
         } else {
             const max_slot = std.math.maxInt(Slot); // TODO agave uses BankForks for this
-            if (shouldDiscardShred(packet, self.root_slot, shred_version, max_slot)) {
+            if (shouldDiscardShred(packet, self.root_slot, self.shred_version, max_slot)) {
                 packet.flags.set(.discard);
             }
         }
     }
 
-    /// Handle a ping message and return
-    fn handlePing(self: *Self, packet: *const Packet, responses: *ArrayList(Packet)) !void {
+    /// Handle a ping message and returns the repair message.
+    fn handlePing(self: *const Self, packet: *const Packet) !?Packet {
         const repair_ping = bincode.readFromSlice(self.allocator, RepairPing, &packet.data, .{}) catch {
             self.metrics.invalid_repair_pings.inc();
-            return;
+            return null;
         };
         const ping = repair_ping.Ping;
         ping.verify() catch {
             self.metrics.invalid_repair_pings.inc();
-            return;
+            return null;
         };
+        const reply: RepairMessage = .{ .Pong = try Pong.init(&ping, self.keypair) };
 
-        const reply = RepairMessage{ .Pong = try Pong.init(&ping, self.keypair) };
-        const reply_packet = try responses.addOne();
+        var reply_packet = Packet.default();
         const reply_bytes = try bincode.writeToSlice(&reply_packet.data, reply, .{});
         reply_packet.size = reply_bytes.len;
         reply_packet.addr = packet.addr;
+        return reply_packet;
     }
 };
 
 fn shouldDiscardShred(
     packet: *const Packet,
     root: Slot,
-    shred_version: u16,
+    shred_version: *const Atomic(u16),
```

Comment:
Another advantage of the packet arrays is that you can reduce the number of atomic loads of this value.

Reply:
Doing more atomic loads is not something we should be worrying about. Not only are they the cheapest atomic memory operation, but on architectures like x86 it's the default order of all loads.